Undo doesn't always scroll to edit location

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hitting undo multiple doesn't always scroll to the edit location. My file is ~485 lines of code, so not particularly long. Sometimes it'll jump to the first site of the last change you're undoing, but then start undoing things in hidden parts of the file without scrolling. 

This can be pretty dangerous as if you start making changes after undoing a bit too far with no visible feedback in the IDE you can very easily lose undo history and break your local codebase.

IMO being able to undo and see what you are undoing **100% of the time** is an absolutely essential behavior. I literally can't continue using Zed if this keeps up.

### Zed Version and System Specs

Zed: v0.166.1 (Zed)
OS: macOS 14.2.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add screenshots or screencasts of the incorrect state / behavior

_No response_

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```

```
<!-- Click above this line and paste or drag-and-drop your log--></details>

Steps to reproduce:

1. Open a file (e.g. cmd-, for settings.json)
2. Make some changes.
3. Trigger `editor: format document` (or save with `format_on_save`)
4. Undo

Check your logs and you will see:
```
[ERROR] No entry in selection_history found for undo. This may correspond to a bug where undo does not update the selection. If this is occurring, please add details to https://github.com/zed-industries/zed/issues/22692
```

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -14218,13 +14218,7 @@ impl Editor {
             }
         };
 
-        let transaction_id_prev = buffer.read_with(cx, |b, cx| {
-            if let Some(singleton) = b.as_singleton() {
-                singleton.read_with(cx, |b, _| b.peek_undo_stack().map(|h| h.transaction_id()))
-            } else {
-                b.last_transaction_id()
-            }
-        });
+        let transaction_id_prev = buffer.read_with(cx, |b, cx| b.last_transaction_id(cx));
         let selections_prev = transaction_id_prev
             .and_then(|transaction_id_prev| {
                 // default to selections as they were after the last edit, if we have them,
@@ -14265,13 +14259,9 @@ impl Editor {
                 })
                 .ok();
 
-            if let Some(transaction_id_now) = buffer.read_with(cx, |b, cx| {
-                if let Some(singleton) = b.as_singleton() {
-                    singleton.read_with(cx, |b, _| b.peek_undo_stack().map(|h| h.transaction_id()))
-                } else {
-                    b.last_transaction_id()
-                }
-            })? {
+            if let Some(transaction_id_now) =
+                buffer.read_with(cx, |b, cx| b.last_transaction_id(cx))?
+            {
                 let has_new_transaction = transaction_id_prev != Some(transaction_id_now);
                 if has_new_transaction {
                     _ = editor.update(cx, |editor, _| {

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -923,8 +923,57 @@ struct SelectionHistory {
     selections_by_transaction:
         HashMap<TransactionId, (Arc<[Selection<Anchor>]>, Option<Arc<[Selection<Anchor>]>>)>,
     mode: SelectionHistoryMode,
-    undo_stack: VecDeque<SelectionHistoryEntry>,
-    redo_stack: VecDeque<SelectionHistoryEntry>,
+    undo_stack: SelectionHistoryRingBuffer,
+    redo_stack: SelectionHistoryRingBuffer,
+}
+
+struct SelectionHistoryRingBuffer {
+    index: usize,
+    entries: [Option<SelectionHistoryEntry>; MAX_SELECTION_HISTORY_LEN],
+}
+
+impl Default for SelectionHistoryRingBuffer {
+    fn default() -> Self {
+        return Self::new();
+    }
+}
+
+impl SelectionHistoryRingBuffer {
+    fn new() -> Self {
+        return Self {
+            index: 0,
+            entries: [0; MAX_SELECTION_HISTORY_LEN].map(|_| None),
+        };
+    }
+
+    fn push(&mut self, entry: SelectionHistoryEntry) {
+        self.entries[self.index].replace(entry);
+        self.index += 1;
+        if self.index >= MAX_SELECTION_HISTORY_LEN {
+            self.index = 0;
+        }
+    }
+
+    fn pop(&mut self) -> Option<SelectionHistoryEntry> {
+        if self.index == 0 {
+            self.index = MAX_SELECTION_HISTORY_LEN;
+        }
+        self.index -= 1;
+        return self.entries[self.index].take();
+    }
+
+    fn clear(&mut self) {
+        self.entries.fill(None);
+    }
+
+    fn back(&self) -> Option<&SelectionHistoryEntry> {
+        let mut index = self.index;
+        if index == 0 {
+            index = MAX_SELECTION_HISTORY_LEN;
+        }
+        index -= 1;
+        return self.entries[index].as_ref();
+    }
 }
 
 impl SelectionHistory {
@@ -972,10 +1021,7 @@ impl SelectionHistory {
             .back()
             .map_or(true, |e| e.selections != entry.selections)
         {
-            self.undo_stack.push_back(entry);
-            if self.undo_stack.len() > MAX_SELECTION_HISTORY_LEN {
-                self.undo_stack.pop_front();
-            }
+            self.undo_stack.push(entry);
         }
     }
 
@@ -985,10 +1031,7 @@ impl SelectionHistory {
             .back()
             .map_or(true, |e| e.selections != entry.selections)
         {
-            self.redo_stack.push_back(entry);
-            if self.redo_stack.len() > MAX_SELECTION_HISTORY_LEN {
-                self.redo_stack.pop_front();
-            }
+            self.redo_stack.push(entry);
         }
     }
 }
@@ -12747,7 +12790,7 @@ impl Editor {
         self.hide_mouse_cursor(&HideMouseCursorOrigin::MovementAction);
         self.end_selection(window, cx);
         self.selection_history.mode = SelectionHistoryMode::Undoing;
-        if let Some(entry) = self.selection_history.undo_stack.pop_back() {
+        if let Some(entry) = self.selection_history.undo_stack.pop() {
             self.change_selections(None, window, cx, |s| {
                 s.select_anchors(entry.selections.to_vec())
             });
@@ -12768,7 +12811,7 @@ impl Editor {
         self.hide_mouse_cursor(&HideMouseCursorOrigin::MovementAction);
         self.end_selection(window, cx);
         self.selection_history.mode = SelectionHistoryMode::Redoing;
-        if let Some(entry) = self.selection_history.redo_stack.pop_back() {
+        if let Some(entry) = self.selection_history.redo_stack.pop() {
             self.change_selections(None, window, cx, |s| {
                 s.select_anchors(entry.selections.to_vec())
             });

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -14175,12 +14175,34 @@ impl Editor {
             }
         };
 
+        let transaction_id_prev = buffer.read_with(cx, |b, cx| {
+            if let Some(singleton) = b.as_singleton() {
+                singleton.read_with(cx, |b, _| b.peek_undo_stack().map(|h| h.transaction_id()))
+            } else {
+                b.last_transaction_id()
+            }
+        });
+        let selections_prev = transaction_id_prev
+            .and_then(|transaction_id_prev| {
+                // default to selections as they were after the last edit, if we have them,
+                // instead of how they are now.
+                // This will make it so that editing, moving somewhere else, formatting, then undoing the format
+                // will take you back to where you made the last edit, instead of staying where you scrolled
+                self.selection_history
+                    .transaction(transaction_id_prev)
+                    .map(|t| t.0.clone())
+            })
+            .unwrap_or_else(|| {
+                log::info!("Failed to determine selections from before format. Falling back to selections when format was initiated");
+                self.selections.disjoint_anchors()
+            });
+
         let mut timeout = cx.background_executor().timer(FORMAT_TIMEOUT).fuse();
         let format = project.update(cx, |project, cx| {
             project.format(buffers, target, true, trigger, cx)
         });
 
-        cx.spawn_in(window, async move |_, cx| {
+        cx.spawn_in(window, async move |editor, cx| {
             let transaction = futures::select_biased! {
                 transaction = format.log_err().fuse() => transaction,
                 () = timeout => {
@@ -14199,6 +14221,23 @@ impl Editor {
                     cx.notify();
                 })
                 .ok();
+
+            if let Some(transaction_id_now) = buffer.read_with(cx, |b, cx| {
+                if let Some(singleton) = b.as_singleton() {
+                    singleton.read_with(cx, |b, _| b.peek_undo_stack().map(|h| h.transaction_id()))
+                } else {
+                    b.last_transaction_id()
+                }
+            })? {
+                let has_new_transaction = transaction_id_prev != Some(transaction_id_now);
+                if has_new_transaction {
+                    _ = editor.update(cx, |editor, _| {
+                        editor
+                            .selection_history
+                            .insert_transaction(transaction_id_now, selections_prev);
+                    });
+                }
+            }
 
             Ok(())
         })

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -923,57 +923,8 @@ struct SelectionHistory {
     selections_by_transaction:
         HashMap<TransactionId, (Arc<[Selection<Anchor>]>, Option<Arc<[Selection<Anchor>]>>)>,
     mode: SelectionHistoryMode,
-    undo_stack: SelectionHistoryRingBuffer,
-    redo_stack: SelectionHistoryRingBuffer,
-}
-
-struct SelectionHistoryRingBuffer {
-    index: usize,
-    entries: [Option<SelectionHistoryEntry>; MAX_SELECTION_HISTORY_LEN],
-}
-
-impl Default for SelectionHistoryRingBuffer {
-    fn default() -> Self {
-        return Self::new();
-    }
-}
-
-impl SelectionHistoryRingBuffer {
-    fn new() -> Self {
-        return Self {
-            index: 0,
-            entries: [0; MAX_SELECTION_HISTORY_LEN].map(|_| None),
-        };
-    }
-
-    fn push(&mut self, entry: SelectionHistoryEntry) {
-        self.entries[self.index].replace(entry);
-        self.index += 1;
-        if self.index >= MAX_SELECTION_HISTORY_LEN {
-            self.index = 0;
-        }
-    }
-
-    fn pop(&mut self) -> Option<SelectionHistoryEntry> {
-        if self.index == 0 {
-            self.index = MAX_SELECTION_HISTORY_LEN;
-        }
-        self.index -= 1;
-        return self.entries[self.index].take();
-    }
-
-    fn clear(&mut self) {
-        self.entries.fill(None);
-    }
-
-    fn back(&self) -> Option<&SelectionHistoryEntry> {
-        let mut index = self.index;
-        if index == 0 {
-            index = MAX_SELECTION_HISTORY_LEN;
-        }
-        index -= 1;
-        return self.entries[index].as_ref();
-    }
+    undo_stack: VecDeque<SelectionHistoryEntry>,
+    redo_stack: VecDeque<SelectionHistoryEntry>,
 }
 
 impl SelectionHistory {
@@ -1021,7 +972,10 @@ impl SelectionHistory {
             .back()
             .map_or(true, |e| e.selections != entry.selections)
         {
-            self.undo_stack.push(entry);
+            self.undo_stack.push_back(entry);
+            if self.undo_stack.len() > MAX_SELECTION_HISTORY_LEN {
+                self.undo_stack.pop_front();
+            }
         }
     }
 
@@ -1031,7 +985,10 @@ impl SelectionHistory {
             .back()
             .map_or(true, |e| e.selections != entry.selections)
         {
-            self.redo_stack.push(entry);
+            self.redo_stack.push_back(entry);
+            if self.redo_stack.len() > MAX_SELECTION_HISTORY_LEN {
+                self.redo_stack.pop_front();
+            }
         }
     }
 }
@@ -12790,7 +12747,7 @@ impl Editor {
         self.hide_mouse_cursor(&HideMouseCursorOrigin::MovementAction);
         self.end_selection(window, cx);
         self.selection_history.mode = SelectionHistoryMode::Undoing;
-        if let Some(entry) = self.selection_history.undo_stack.pop() {
+        if let Some(entry) = self.selection_history.undo_stack.pop_back() {
             self.change_selections(None, window, cx, |s| {
                 s.select_anchors(entry.selections.to_vec())
             });
@@ -12811,7 +12768,7 @@ impl Editor {
         self.hide_mouse_cursor(&HideMouseCursorOrigin::MovementAction);
         self.end_selection(window, cx);
         self.selection_history.mode = SelectionHistoryMode::Redoing;
-        if let Some(entry) = self.selection_history.redo_stack.pop() {
+        if let Some(entry) = self.selection_history.redo_stack.pop_back() {
             self.change_selections(None, window, cx, |s| {
                 s.select_anchors(entry.selections.to_vec())
             });

--- a/crates/editor/src/editor_tests.rs
+++ b/crates/editor/src/editor_tests.rs
@@ -5874,6 +5874,83 @@ async fn test_select_all_matches_does_not_scroll(cx: &mut TestAppContext) {
 }
 
 #[gpui::test]
+async fn test_undo_format_scrolls_to_last_edit_pos(cx: &mut TestAppContext) {
+    init_test(cx, |_| {});
+
+    let mut cx = EditorLspTestContext::new_rust(
+        lsp::ServerCapabilities {
+            document_formatting_provider: Some(lsp::OneOf::Left(true)),
+            ..Default::default()
+        },
+        cx,
+    )
+    .await;
+
+    cx.set_state(indoc! {"
+        line 1
+        line 2
+        linˇe 3
+        line 4
+        line 5
+    "});
+
+    // Make an edit
+    cx.update_editor(|editor, window, cx| {
+        editor.handle_input("X", window, cx);
+    });
+
+    // Move cursor to a different position
+    cx.update_editor(|editor, window, cx| {
+        editor.change_selections(None, window, cx, |s| {
+            s.select_ranges([Point::new(4, 2)..Point::new(4, 2)]);
+        });
+    });
+
+    cx.assert_editor_state(indoc! {"
+        line 1
+        line 2
+        linXe 3
+        line 4
+        liˇne 5
+    "});
+
+    cx.lsp
+        .set_request_handler::<lsp::request::Formatting, _, _>(move |_, _| async move {
+            Ok(Some(vec![lsp::TextEdit::new(
+                lsp::Range::new(lsp::Position::new(0, 0), lsp::Position::new(0, 0)),
+                "PREFIX ".to_string(),
+            )]))
+        });
+
+    cx.update_editor(|editor, window, cx| editor.format(&Default::default(), window, cx))
+        .unwrap()
+        .await
+        .unwrap();
+
+    cx.assert_editor_state(indoc! {"
+        PREFIX line 1
+        line 2
+        linXe 3
+        line 4
+        liˇne 5
+    "});
+
+    // Undo formatting
+    cx.update_editor(|editor, window, cx| {
+        editor.undo(&Default::default(), window, cx);
+    });
+
+    // Verify cursor moved back to position after edit
+    cx.assert_editor_state(indoc! {"
+        line 1
+        line 2
+        linXˇe 3
+        line 4
+        line 5
+    "});
+}
+
+#[gpui::test]
 async fn test_select_next_with_multiple_carets(cx: &mut TestAppContext) {
     init_test(cx, |_| {});
 

--- a/crates/multi_buffer/src/multi_buffer.rs
+++ b/crates/multi_buffer/src/multi_buffer.rs
@@ -1118,9 +1118,16 @@ impl MultiBuffer {
         self.history.start_transaction(now)
     }
 
-    pub fn last_transaction_id(&self) -> Option<TransactionId> {
-        let last_transaction = self.history.undo_stack.last()?;
-        return Some(last_transaction.id);
+    pub fn last_transaction_id(&self, cx: &App) -> Option<TransactionId> {
+        if let Some(buffer) = self.as_singleton() {
+            return buffer.read_with(cx, |b, _| {
+                b.peek_undo_stack()
+                    .map(|history_entry| history_entry.transaction_id())
+            });
+        } else {
+            let last_transaction = self.history.undo_stack.last()?;
+            return Some(last_transaction.id);
+        }
     }
 
     pub fn end_transaction(&mut self, cx: &mut Context<Self>) -> Option<TransactionId> {

